File: LeetCode/remote/search_a_2d_matrix.cs
//  https://leetcode.com/problems/search-a-2d-matrix-ii/
//
//  Write an efficient algorithm that searches for a value in an m x n matrix. This matrix has the following properties:
//
//      Integers in each row are sorted in ascending from left to right.
//      Integers in each column are sorted in ascending from top to bottom.

using System;

public class Solution {
    //  
    //  Submission Details
    //  127 / 127 test cases passed.
    //      Status: Accepted
    //      Runtime: 304 ms
    //          
    //          Submitted: 0 minutes ago
    //
    //          https://leetcode.com/submissions/detail/49418465/
    //
    //          Idea was good but I kinda got lost in the details
    public bool SearchMatrix(int[,] matrix, int target) {
        var line = 0;
        var column = matrix.GetLength(1) - 1;
        while (line < matrix.GetLength(0) && column >= 0)
        {
            if (matrix[line, column] == target)
            {
                return true;
            }

            if (matrix[line, column] > target)
            {
                column--;
            }
            else
            {
                line++;
            }
        }

        return false;
    }

    static void Main()
    {
        var matrix = new [,] {
                      {1,   4,  7, 11, 15},
                      {2,   5,  8, 12, 19},
                      {3,   6,  9, 16, 22},
                      {10, 13, 14, 17, 24},
                      {18, 21, 23, 26, 30}
                    };
        Console.WriteLine(new Solution().SearchMatrix(matrix, 5));
        Console.WriteLine(new Solution().SearchMatrix(matrix, 20));
    }
}

File: LeetCode/remote/search_a_2d_matrix.rb
#   https://leetcode.com/problems/search-a-2d-matrix/
#
#   Write an efficient algorithm that searches for a value in an m x n matrix. This matrix has the following properties:
#
#   Integers in each row are sorted from left to right.
#   The first integer of each row is greater than the last integer of the previous row.
#   https://leetcode.com/submissions/detail/53871278/
#
#   Submission Details
#   134 / 134 test cases passed.
#       Status: Accepted
#       Runtime: 72 ms
#           
#           Submitted: 0 minutes ago
# @param {Integer[][]} matrix
# @param {Integer} target
# @return {Boolean}
def search_matrix(matrix, target)
    row = 0
    column = matrix[0].size - 1
    while row < matrix.size && column >=0
        return true if matrix[row][column] == target
        if matrix[row][column] > target
            column -= 1
        else
            row += 1
        end
    end
    
    return false
end

